File: src/main/java/org/compiere/dbPort/Join.java
package org.compiere.dbPort;

import org.idempiere.common.util.Util;

/**
 * Join Clause.
 *
 * <pre>
 *  f.AD_Column_ID = c.AD_Column_ID(+)
 *  </pre>
 *
 * @author Jorg Janke
 * @version $Id: Join.java,v 1.2 2006/07/30 00:55:04 jjanke Exp $
 */
public class Join {
    private String m_joinClause;
    private String m_mainTable;
    private String m_mainAlias;
    private String m_joinTable;
    private String m_joinAlias;
    private boolean m_left;
    private String m_condition;
    /**
     * Constructor
     *
     * @param joinClause
     */
    public Join(String joinClause) {
        if (joinClause == null) throw new IllegalArgumentException("Join - clause cannot be null");
        evaluate(joinClause);
    } //  Join

    /**
     * Evaluate the clause. e.g. tb.AD_User_ID(+)=? f.AD_Column_ID = c.AD_Column_ID(+)
     *
     * @param joinClause
     */
    private void evaluate(String joinClause) {
        m_joinClause = joinClause;
        int indexEqual = joinClause.indexOf('=');
        m_left = indexEqual < joinClause.indexOf("(+)"); //  converts to LEFT if true
        //  get table alias of it
        if (m_left) //  f.AD_Column_ID = c.AD_Column_ID(+)  => f / c
        {
            m_mainAlias = joinClause.substring(0, Util.findIndexOf(joinClause, '.', '=')).trim(); //  f
            int end = joinClause.indexOf('.', indexEqual);
            if (end == -1) //  no alias
                end = joinClause.indexOf('(', indexEqual);
            m_joinAlias = joinClause.substring(indexEqual + 1, end).trim(); //  c
        } else //  f.AD_Column_ID(+) = c.AD_Column_ID  => c / f
        {
            int end = joinClause.indexOf('.', indexEqual);
            if (end == -1) //  no alias
                end = joinClause.length();
            m_mainAlias = joinClause.substring(indexEqual + 1, end).trim(); //  c
            m_joinAlias = joinClause.substring(0, Util.findIndexOf(joinClause, '.', '(')).trim(); //  f
        }
        m_condition = Util.replace(joinClause, "(+)", "").trim();
    } //  evaluate

    /**
     * Get origial Join Clause. e.g. f.AD_Column_ID = c.AD_Column_ID(+)
     *
     * @return Join cluase
     */
    public String getJoinClause() {
        return m_joinClause;
    } //  getJoinClause

    /**
     * Get Main Table Alias
     *
     * @return Main Table Alias
     */
    public String getMainAlias() {
        return m_mainAlias;
    } //  getMainAlias

    /**
     * Get Join Table Alias
     *
     * @return Join Table Alias
     */
    public String getJoinAlias() {
        return m_joinAlias;
    } //  getJoinAlias

    /**
     * Is Left Aouter Join
     *
     * @return true if left outer join
     */
    public boolean isLeft() {
        return m_left;
    } //  isLeft

    /**
     * Get Join condition. e.g. f.AD_Column_ID = c.AD_Column_ID
     *
     * @return join condition
     */
    public String getCondition() {
        return m_condition;
    } //  getCondition

    /** ********************************************************************** */

    /**
     * Get Main Table Name
     *
     * @return Main Table Name
     */
    public String getMainTable() {
        return m_mainTable;
    } //  getMainTable

    /**
     * Set Main Table Name. If table name equals alias, the alias is set to ""
     *
     * @param mainTable
     */
    public void setMainTable(String mainTable) {
        if (mainTable == null || mainTable.length() == 0) return;
        m_mainTable = mainTable;
        if (m_mainAlias.equals(mainTable)) m_mainAlias = "";
    } //  setMainTable

    /**
     * Get Join Table Name
     *
     * @return Join Table Name
     */
    public String getJoinTable() {
        return m_joinTable;
    } //  getJoinTable

    /**
     * Set Main Table Name. If table name equals alias, the alias is set to ""
     *
     * @param joinTable
     */
    public void setJoinTable(String joinTable) {
        if (joinTable == null || joinTable.length() == 0) return;
        m_joinTable = joinTable;
        if (m_joinAlias.equals(joinTable)) m_joinAlias = "";
    } //  setJoinTable

    /** ********************************************************************** */

    /**
     * This Join is a condition of the first Join. e.g. tb.AD_User_ID(+)=? or tb.AD_User_ID(+)='123'
     *
     * @param first
     * @return true if condition
     */
    public boolean isConditionOf(Join first) {
        //  same main table
        return m_mainTable == null //  did not find Table from "Alias"
                && (first.getJoinTable().equals(m_joinTable) //  same join table
                || first.getMainAlias().equals(m_joinTable));
    } //  isConditionOf

    /**
     * String representation
     *
     * @return info
     */
    public String toString() {
        StringBuilder sb = new StringBuilder("Join[");
        sb.append(m_joinClause)
                .append(" - Main=")
                .append(m_mainTable)
                .append("/")
                .append(m_mainAlias)
                .append(", Join=")
                .append(m_joinTable)
                .append("/")
                .append(m_joinAlias)
                .append(", Left=")
                .append(m_left)
                .append(", Condition=")
                .append(m_condition)
                .append("]");
        return sb.toString();
    } //  toString
} //  Join

File: src/main/kotlin/org/compiere/dbPort/Join.kt
package org.compiere.dbPort

import org.idempiere.common.util.Util

/**
 * Join Clause.
 *
 * <pre>
 * f.AD_Column_ID = c.AD_ColumnId(+)
</pre> *
 *
 * @author Jorg Janke
 * @version $Id: Join.java,v 1.2 2006/07/30 00:55:04 jjanke Exp $
 */
class Join
/**
 * Constructor
 *
 * @param joinClause
 */
    (joinClause: String?) {
    private var m_joinClause: String? = null
    /**
     * Get Main Table Name
     *
     * @return Main Table Name
     */
    /**
     * Set Main Table Name. If table name equals alias, the alias is set to ""
     *
     * @param mainTable
     */
    var mainTable: String? = null
        set(mainTable) {
            if (mainTable == null || //  did not find Table from "Alias"
                mainTable.isEmpty()
            ) return
            field = mainTable
            if (mainAlias == mainTable) mainAlias = ""
        } //  getMainTable
    //  setMainTable
    /**
     * Get Main Table Alias
     *
     * @return Main Table Alias
     */
    var mainAlias: String? = null
        private set //  getMainAlias
    /**
     * Get Join Table Name
     *
     * @return Join Table Name
     */
    /**
     * Set Main Table Name. If table name equals alias, the alias is set to ""
     *
     * @param joinTable
     */
    var joinTable: String? = null
        set(joinTable) {
            if (joinTable == null || joinTable.isEmpty()) return
            field = joinTable
            if (joinAlias == joinTable) joinAlias = ""
        } //  getJoinTable
    //  setJoinTable
    /**
     * Get Join Table Alias
     *
     * @return Join Table Alias
     */
    var joinAlias: String? = null
        private set //  getJoinAlias
    /**
     * Is Left Aouter Join
     *
     * @return true if left outer join
     */
    var isLeft: Boolean = false
        private set //  isLeft
    /**
     * Get Join condition. e.g. f.AD_Column_ID = c.AD_Column_ID
     *
     * @return join condition
     */
    var condition: String? = null
        private set //  getCondition

    init {
        if (joinClause == null) throw IllegalArgumentException("Join - clause cannot be null")
        evaluate(joinClause)
    } //  Join

    /**
     * Evaluate the clause. e.g. tb.AD_UserId(+)=? f.AD_Column_ID = c.AD_ColumnId(+)
     *
     * @param joinClause
     */
    private fun evaluate(joinClause: String) {
        m_joinClause = joinClause
        val indexEqual = joinClause.indexOf('=')
        isLeft = indexEqual < joinClause.indexOf("(+)") //  converts to LEFT if true
        //  get table alias of it
        if (isLeft)
        //  f.AD_Column_ID = c.AD_ColumnId(+)  => f / c
        {
            mainAlias = joinClause.substring(0, Util.findIndexOf(joinClause, '.', '=')).trim { it <= ' ' } //  f
            var end = joinClause.indexOf('.', indexEqual)
            if (end == -1)
            //  no alias
                end = joinClause.indexOf('(', indexEqual)
            joinAlias = joinClause.substring(indexEqual + 1, end).trim { it <= ' ' } //  c
        } else
        //  f.AD_ColumnId(+) = c.AD_Column_ID  => c / f
        {
            var end = joinClause.indexOf('.', indexEqual)
            if (end == -1)
            //  no alias
                end = joinClause.length
            mainAlias = joinClause.substring(indexEqual + 1, end).trim { it <= ' ' } //  c
            joinAlias = joinClause.substring(0, Util.findIndexOf(joinClause, '.', '(')).trim { it <= ' ' } //  f
        }
        condition = Util.replace(joinClause, "(+)", "").trim { it <= ' ' }
    } //  evaluate

    /**
     * This Join is a condition of the first Join. e.g. tb.AD_UserId(+)=? or tb.AD_UserId(+)='123'
     *
     * @param first
     * @return true if condition
     */
    fun isConditionOf(first: Join): Boolean {
        //  same main table
        return mainTable == null && (first.joinTable == joinTable || //  same join table
                first.mainAlias == joinTable)
    } //  isConditionOf

    /**
     * String representation
     *
     * @return info
     */
    override fun toString(): String {
        return "Join[" + m_joinClause +
                " - Main=" +
                mainTable +
                "/" +
                mainAlias +
                ", Join=" +
                joinTable +
                "/" +
                joinAlias +
                ", Left=" +
                isLeft +
                ", Condition=" +
                condition +
                "]"
    } //  toString
} //  Join
